### PR TITLE
PLG-64 Pagination for all resources during m2 import

### PR DESCRIPTION
**Pull request by nedick007**
Initial commit of new AJAX import logic.

### CODE REVIEW

File: Model/CategoryData.php [hunk 1]
```diff
@@ -12,17 +12,24 @@ public function __construct(
         $this->storeManager = $storeManager;
     }
 
-    public function getCategories($storeId)
+    public function getCategoryQuery($storeId = 0)
```
Comment (adamof) [inline Model/CategoryData.php]: Why have you set this default value here?

File: Model/CustomerData.php [hunk 2]
```diff
@@ -12,20 +12,35 @@ public function __construct(
...
     }
 
-    public function getCustomers($storeId)
+//    private $chunkItems = 1;
+//
+//    public function getCustomerChunks($storeId = 0)
+//    {
+//        $totalCustomers = $this->getCustomerQuery($storeId)->getSize();
+//        return (int) ceil($totalCustomers / $this->chunkItems);
+//    }
```
Comment (adamof) [inline Model/CustomerData.php]: remove the commented code

File: Model/CustomerData.php [hunk 3]
```diff
@@ -12,20 +12,35 @@ public function __construct(
...
+
+    public function getCustomerQuery($storeId = 0)
```
Comment (adamof) [inline Model/CustomerData.php]: same ?

File: Model/Import.php [hunk 4]
```diff
@@ -14,46 +14,46 @@ class Import
...
+        $this->productData     = $productData;
+        $this->orderData       = $orderData;
+        $this->resolver        = $resolver;
     }
 
-    /**
-     * @return int
-     */
-    public function getTotalChunks()
+    public function getCustomerChunks($storeId = 0)
```
Comment (adamof) [inline Model/Import.php]: same Q

File: Model/Import.php [hunk 5]
```diff
@@ -14,46 +14,46 @@ class Import
...
     }
 
-    /**
-     * Get chunk orders
-     *
-     * @param  int
-     * @return
-     */
-    public function getOrders($storeId, $chunkId)
+    public function getCategoryChunks($storeId = 0)
```
Comment (adamof) [inline Model/Import.php]: same Q

File: Model/Import.php [hunk 6]
```diff
@@ -14,46 +14,46 @@ class Import
...
+    {
+        $totalCategories = $this->categoryData->getCategoryQuery($storeId)->getSize();
+        return (int) ceil($totalCategories / $this->chunkItems);
+    }
+    
+    public function getProductChunks($storeId = 0)
```
Comment (adamof) [inline Model/Import.php]: same q

File: Model/OrderData.php [hunk 7]
```diff
@@ -63,7 +65,7 @@ public function getOrders($storeId)
         return $ordersArray;
     }
 
-    protected function getOrderQuery($storeId = 0)
+    public function getOrderQuery($storeId = 0)
```
Comment (adamof) [inline Model/OrderData.php]: same Q

File: Controller/Adminhtml/Import/Ajax.php [hunk 8]
```diff
@@ -47,46 +39,55 @@ public function __construct(
...
-            $orders = $this->import->getOrders($storeId, $chunkId);
-            // Send orders via API helper method
-            $this->helper->callBatchApi($storeId, $orders, false);
-            $result['success'] = true;
+            // file_put_contents(__DIR__ . 'Request.log', $client->customerBatch($this->import->customerData->getCustomers($storeId, $chunkId, $this->import->chunkItems)), FILE_APPEND); // Used for debug purposes;
 
-            if ($chunkId == $totalChunks - 1) {
-                $this->helper->createActivity($storeId, 'import_end');
+            switch ($importStatus) {
+                case 1:
```
Comment (adamof) [inline Controller/Adminhtml/Import/Ajax.php]: why don't you use strings to make this more understandable?

File: view/adminhtml/web/js/import.js [hunk 9]
```diff
@@ -42,49 +46,96 @@ define([
...
-                            self.updateImportingMessage("<span style='color: green;'>" + $t('Done! Please expect up to 30 minutes for your historical data to appear in Metrilo.') + "</span>");
-                            self.element.removeClass('disabled').addClass('success').text($t('Orders imported'));
-                        }
-                    } else {
-                        self.updateImportingMessage("<span style='color: red;'>" + response.message + "</span>");
+
+                self.ajaxPostWithRetry(self.options.submitUrl, data, 3, function(response) {
+                    var newChunkId = chunkId + 1;
+                    switch (importStatus) {
+                        case 1:
```
Comment (adamof) [inline view/adminhtml/web/js/import.js]: same question, why not use strings?

File: view/adminhtml/web/js/import.js [hunk 10]
```diff
@@ -42,49 +46,96 @@ define([
...
+                    setTimeout(function() {
+                        self.chunkSync(0, self.options.importStatus);
+                    }, 2000);
+                }
+            },
+
+            ajaxPostWithRetry: function(url, data, retryCount, callback) {
+                if(retryCount) {
+                    $.post(url, data, function(response) {
+                        console.log('Successfull import of chunkData');
```
Comment (adamof) [inline view/adminhtml/web/js/import.js]: probably don't need those

**Comment by adamof** [comment]
@tabaliov, please review the JS part. This needs to be rock-solid

File: Controller/Adminhtml/Import/Ajax.php (hunk 8 above)
Comment (nedick007) [reply Controller/Adminhtml/Import/Ajax.php]: Statuses changed customer, category, product, order.

File: Model/CategoryData.php (hunk 1 above)
Comment (nedick007) [reply Model/CategoryData.php]: Agree that it is unnecessary. Removed.

File: Model/CustomerData.php (hunk 2 above)
Comment (nedick007) [reply Model/CustomerData.php]: Removed.

File: Model/CustomerData.php (hunk 3 above)
Comment (nedick007) [reply Model/CustomerData.php]: Removed the default value.

File: Model/Import.php (hunk 4 above)
Comment (nedick007) [reply Model/Import.php]: Removed the default value.

File: Model/Import.php (hunk 5 above)
Comment (nedick007) [reply Model/Import.php]: Removed the default value.

File: Model/Import.php (hunk 6 above)
Comment (nedick007) [reply Model/Import.php]: Removed the default value.

File: Model/OrderData.php (hunk 7 above)
Comment (nedick007) [reply Model/OrderData.php]: Removed the default value.

File: view/adminhtml/web/js/import.js (hunk 9 above)
Comment (nedick007) [reply view/adminhtml/web/js/import.js]: Edited importStatus to strings of corresponding import type.

File: view/adminhtml/web/js/import.js (hunk 10 above)
Comment (nedick007) [reply view/adminhtml/web/js/import.js]: It was used for debug purposes. Removed.

**Comment by nedick007** [comment]
@tabaliov requested changes from @adamof are uploaded.

**Comment by nedick007** [comment]
@tabaliov the js modifications are in 'PLG-64: retry import of failed chunk modification' commit. Rest is for @adamof

File: Model/CategoryData.php [hunk 11]
```diff
@@ -12,17 +12,24 @@ public function __construct(
...
+        return $this->categoryCollection->create()->addAttributeToSelect('name')
                     ->joinTable(
                         ['url' => 'url_rewrite'],
                         'entity_id = entity_id',
                         ['request_path', 'store_id'],
                         ['entity_type' => 'category', 'store_id' => $storeId]
                     );
+    }
+
+    public function getCategories($storeId, $chunkId, $chunkItems)
```
Comment (adamof) [inline Model/CategoryData.php]: This comment is about all of the `Data` classes - product, order, category, customer:
Is the `$chunkItems` parameter ever different?
If not - why do you pass it half of the time and not just get it as a constant from the Import class in the `getXXX` function?

File: view/adminhtml/web/js/import.js [hunk 12]
```diff
@@ -42,49 +46,103 @@ define([
...
 
             /**
-             * Post chunk id to proceed orders to Metrilo
+             * Post chunk id to proceed data to Metrilo
              *
              * @param  {integer} chunkId
              * @return {void}
              */
-            chunkSync: function(chunkId) {
+            chunkSync: function(chunkId, importStatus, retry) {
```
Comment (adamof) [inline view/adminhtml/web/js/import.js]: What is `importStatus`?

Comment (adamof) [reply view/adminhtml/web/js/import.js]: What is `retry`?

File: view/adminhtml/web/js/import.js [hunk 13]
```diff
@@ -42,49 +46,103 @@ define([
...
+                                self.element.removeClass('disabled').addClass('success').text($t('Finished Import.'));
+                            }
+                            break;
+                        default:
+                            return false;
                     }
                 });
             },
 
+            chunkType: function(newChunkId, current, next) {
```
Comment (adamof) [inline view/adminhtml/web/js/import.js]: What does `chunkType` function do? The name is quite obscure

File: view/adminhtml/web/js/import.js [hunk 14]
```diff
@@ -42,49 +46,103 @@ define([
...
+                            break;
+                        case 'category':
+                            self.chunkType(newChunkId, 'category', 'product');
+                            break;
+                        case 'product':
+                            self.chunkType(newChunkId, 'product', 'order');
+                            break;
+                        case 'order':
+                            if(newChunkId < self.options.orderChunks) {
+                                setTimeout(function() {
```
Comment (adamof) [inline view/adminhtml/web/js/import.js]: Why do you need this `setTimeout`?

File: view/adminhtml/web/js/import.js [hunk 15]
```diff
@@ -42,49 +46,103 @@ define([
...
+                        case 'customer':
+                            self.chunkType(newChunkId, 'customer', 'category');
+                            break;
+                        case 'category':
+                            self.chunkType(newChunkId, 'category', 'product');
+                            break;
+                        case 'product':
+                            self.chunkType(newChunkId, 'product', 'order');
+                            break;
+                        case 'order':
```
Comment (adamof) [inline view/adminhtml/web/js/import.js]: Why is the `order` handling so different?

File: view/adminhtml/web/js/import.js [hunk 16]
```diff
@@ -42,49 +46,103 @@ define([
...
                 });
             },
 
+            chunkType: function(newChunkId, current, next) {
+                var self = this;
+                if (self.options[`${current}Chunks`] > 0) {
+                    self.options.percentage = (100 / self.options[`${current}Chunks`]);
+                }
+
+                if(newChunkId < self.options[`${current}Chunks`]) {
```
Comment (adamof) [inline view/adminhtml/web/js/import.js]: This logic seems duplicated the switch case above for orders

File: view/adminhtml/web/js/import.js [hunk 17]
```diff
@@ -42,49 +46,103 @@ define([
...
+            },
+
+            ajaxPostWithRetry: function(url, data, retryCount, callback) {
+                self = this;
+                if(retryCount) {
+                    $.post(url, data, function(response) {
+                        callback(response);
+                    }).fail(function () {
+                        console.log('fail!', data, retryCount);
+                        setTimeout(function() {
```
Comment (adamof) [inline view/adminhtml/web/js/import.js]: why the timeout?

File: view/adminhtml/web/js/import.js (hunk 12 above)
Comment (nedick007) [reply view/adminhtml/web/js/import.js]: Renamed the parameters to importStatus = importType (marking the current importing type)
and retry = retryStatus (true if there was an error during the specific chunk import, triggering retry of the whole importing step)

**Comment by nedick007** [comment]
@adamof bump review

**Comment by nedick007** [comment]
@adamof bump

**Pull request merged**